### PR TITLE
Install Kyma on Owned Cluster

### PR DESCRIPTION
**Pull request by ralikio**
**Description**

Changes proposed in this pull request:

- Base functionality for provisioning without cluster creation.

**Related issue(s)**

### CODE REVIEW

File: components/kyma-environment-broker/internal/process/deprovisioning/remove_runtime.go [hunk 1]
```diff
@@ -34,6 +35,7 @@ func (s *RemoveRuntimeStep) Name() string {
 }
 
 func (s *RemoveRuntimeStep) Run(operation internal.DeprovisioningOperation, log logrus.FieldLogger) (internal.DeprovisioningOperation, time.Duration, error) {
+
```
Comment (tillknuesting) [inline components/kyma-environment-broker/internal/process/deprovisioning/remove_runtime.go:38]:
```suggestion
```

File: components/kyma-environment-broker/internal/process/input/builder.go [hunk 2]
```diff
@@ -141,7 +141,10 @@ func (f *InputBuilderFactory) getHyperscalerProviderForPlanID(planID string, pla
 		provider = &cloudProvider.AWSInput{
 			MultiZone: f.config.MultiZoneCluster,
 		}
+	case broker.OwnClusterPlanID:
+		provider = &cloudProvider.NoHyperscalerInput{}
 		// insert cases for other providers like AWS or GCP
+
```
Comment (tillknuesting) [inline components/kyma-environment-broker/internal/process/input/builder.go:147]:
```suggestion
```

File: components/kyma-environment-broker/internal/provider/owncluster_provider.go [hunk 3]
```diff
@@ -0,0 +1,27 @@
+package provider
+
+import (
+	"github.com/kyma-project/control-plane/components/kyma-environment-broker/internal"
+	"github.com/kyma-project/control-plane/components/provisioner/pkg/gqlschema"
+)
+
+const ()
```
Comment (tillknuesting) [inline components/kyma-environment-broker/internal/provider/owncluster_provider.go:8]:
```suggestion
```

File: components/kyma-environment-broker/internal/provider/owncluster_provider.go [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+
+type NoHyperscalerInput struct {
+}
+
+func (p *NoHyperscalerInput) Defaults() *gqlschema.ClusterConfigInput {
+	return &gqlschema.ClusterConfigInput{}
+}
+
+func (p *NoHyperscalerInput) ApplyParameters(input *gqlschema.ClusterConfigInput, pp internal.ProvisioningParameters) {
+	// TODO: validate and moge kubeconfit here?
```
Comment (tillknuesting) [inline components/kyma-environment-broker/internal/provider/owncluster_provider.go:18]:
```suggestion
	// TODO: validate and merge kubeconfig here?
```

File: components/kyma-environment-broker/internal/broker/instance_create.go [hunk 5]
```diff
@@ -139,8 +140,8 @@ func (b *ProvisionEndpoint) Provision(ctx context.Context, instanceID string, de
 		return b.handleExistingOperation(existingOperation, provisioningParameters)
 	}
 
-	// create SKR shoot name
 	shootName := gardener.CreateShootName()
+	shootDomain := strings.Trim(b.shootDomain, ".")
```
Comment (tillknuesting) [inline components/kyma-environment-broker/internal/broker/instance_create.go:144]: would url.Parse fit here to process a URl?

File: components/kyma-environment-broker/cmd/broker/main.go [hunk 6]
```diff
@@ -947,3 +951,7 @@ func NewClusterOrchestrationProcessingQueue(ctx context.Context, db storage.Brok
 
 	return queue
 }
+
+func noKubeConfig(operation internal.ProvisioningOperation) bool {
```
Comment (piotrmiskiewicz) [inline components/kyma-environment-broker/cmd/broker/main.go:955]: change the name of the function to ownClusterPlan or something like this

**Comment by piotrmiskiewicz** [comment]
General questions, issue we should think about:
 - kubeconfig rotation
 - can we extract shootdomain from the kubeconfig
 - should we check (and how) if the cluster is big enough for the Kyma
 - are we planning to use clusters from one gardener project (only shoot name would be necessary)

File: components/kyma-environment-broker/cmd/broker/main.go [hunk 7]
```diff
@@ -701,8 +704,9 @@ func NewProvisioningProcessingQueue(ctx context.Context, provisionManager *proce
 			step:  provisioning.NewExternalEvalStep(externalEvalCreator),
 		},
 		{
-			stage: postActionsStageName,
-			step:  provisioning.NewRuntimeTagsStep(internalEvalUpdater, provisionerClient),
+			stage:     postActionsStageName,
+			step:      provisioning.NewRuntimeTagsStep(internalEvalUpdater, provisionerClient),
```
Comment (piotrmiskiewicz) [inline components/kyma-environment-broker/cmd/broker/main.go:708]: should we skip AvS for that?

File: components/kyma-environment-broker/internal/process/input/builder.go [hunk 8]
```diff
@@ -245,6 +248,11 @@ func (f *InputBuilderFactory) initProvisionRuntimeInput(provider HyperscalerInpu
 		},
 	}
 
+	// TODO: omit this section
```
Comment (piotrmiskiewicz) [inline components/kyma-environment-broker/internal/process/input/builder.go:251]: why do we need this section/

File: components/kyma-environment-broker/internal/process/input/builder.go [hunk 9]
```diff
@@ -227,7 +230,7 @@ func (f *InputBuilderFactory) provideComponentList(version internal.RuntimeVersi
 	return mapToGQLComponentConfigurationInput(allComponents), nil
 }
 
-func (f *InputBuilderFactory) initProvisionRuntimeInput(provider HyperscalerInputProvider, version internal.RuntimeVersionData, config *internal.ConfigForPlan) (gqlschema.ProvisionRuntimeInput, error) {
+func (f *InputBuilderFactory) initProvisionRuntimeInput(pp internal.ProvisioningParameters, provider HyperscalerInputProvider, version internal.RuntimeVersionData, config *internal.ConfigForPlan) (gqlschema.ProvisionRuntimeInput, error) {
```
Comment (piotrmiskiewicz) [inline components/kyma-environment-broker/internal/process/input/builder.go:233]: it looks like pp is not used

File: components/kyma-environment-broker/internal/process/provisioning/create_runtime_without_kyma_step.go [hunk 10]
```diff
@@ -59,28 +61,43 @@ func (s *CreateRuntimeWithoutKymaStep) Run(operation internal.Operation, log log
...
-	switch {
-	case kebError.IsTemporaryError(err):
-		log.Errorf("call to provisioner failed (temporary error): %s", err)
-		return operation, 5 * time.Second, nil
-	case err != nil:
-		log.Errorf("call to Provisioner failed: %s", err)
-		return s.operationManager.OperationFailed(operation, "call to the provisioner service failed", err, log)
+	var runtimeId string
+	var provisionerOperationId string
+	if operation.ProvisioningParameters.PlanID == broker.OwnClusterPlanID {
```
Comment (piotrmiskiewicz) [inline components/kyma-environment-broker/internal/process/provisioning/create_runtime_without_kyma_step.go:74]: I don't like the fact that step "create runtime without kyma" is not creating runtime for special plan. I'd prefer to not run such step in such condition

File: components/kyma-environment-broker/internal/process/provisioning/get_kubeconfig_step.go [hunk 11]
```diff
@@ -36,6 +37,22 @@ func (s *GetKubeconfigStep) Run(operation internal.Operation, log logrus.FieldLo
 	if operation.Kubeconfig != "" {
 		return operation, 0, nil
 	}
+
+	if operation.ProvisioningParameters.PlanID == broker.OwnClusterPlanID {
+		operation.Kubeconfig = operation.ProvisioningParameters.Parameters.Kubeconfig
```
Comment (piotrmiskiewicz) [inline components/kyma-environment-broker/internal/process/provisioning/get_kubeconfig_step.go:42]: is the kubeconfig field persisted? if yes, is it encrypted?

File: components/kyma-environment-broker/internal/provider/owncluster_provider.go [hunk 12]
```diff
@@ -0,0 +1,27 @@
...
+}
+
+func (p *NoHyperscalerInput) Profile() gqlschema.KymaProfile {
+	return gqlschema.KymaProfileEvaluation
```
Comment (piotrmiskiewicz) [inline components/kyma-environment-broker/internal/provider/owncluster_provider.go:22]: should we use evaluation?

File: components/kyma-environment-broker/cmd/broker/main.go (hunk 7 above)
Comment (piotrmiskiewicz) [reply components/kyma-environment-broker/cmd/broker/main.go:708]: @ebensom can we skip AvS tags for "bring your own gardener cluster"

File: components/kyma-environment-broker/internal/broker/instance_create.go [hunk 13]
```diff
@@ -153,8 +154,60 @@ func (b *ProvisionEndpoint) Provision(ctx context.Context, instanceID string, de
 		logger.Errorf("cannot create new operation: %s", err)
 		return domain.ProvisionedServiceSpec{}, errors.New("cannot create new operation")
 	}
+
```
Comment (piotrmiskiewicz) [inline components/kyma-environment-broker/internal/broker/instance_create.go:157]: let's reorganize this block of code by extracting to separate method or something like this (let's discuss)

File: components/kyma-environment-broker/internal/broker/plans.go [hunk 14]
```diff
@@ -285,6 +290,7 @@ func Plans(plans PlansConfig, provider internal.CloudProvider, includeAdditional
 		AzureLitePlanID: defaultServicePlan(AzureLitePlanID, AzureLitePlanName, plans, azureLiteSchema, AzureLiteSchema(azureLiteMachinesDisplay, azureLiteMachines, includeAdditionalParamsInSchema, true)),
 		FreemiumPlanID:  defaultServicePlan(FreemiumPlanID, FreemiumPlanName, plans, freemiumSchema, FreemiumSchema(provider, includeAdditionalParamsInSchema, true)),
 		TrialPlanID:     defaultServicePlan(TrialPlanID, TrialPlanName, plans, trialSchema, TrialSchema(includeAdditionalParamsInSchema, true)),
+    OwnClusterPlanID: defaultServicePlan(OwnClusterPlanID, OwnClusterPlanName, plans, trialSchema, TrialSchema(includeAdditionalParamsInSchema, true)),
```
Comment (piotrmiskiewicz) [inline components/kyma-environment-broker/internal/broker/plans.go:293]: we cannot use trial schema which contains one field. We need kubeconfig/shootName/shootDomain there

File: components/kyma-environment-broker/internal/process/provisioning/create_cluster_configuration.go [hunk 15]
```diff
@@ -40,6 +40,8 @@ func (s *CreateClusterConfigurationStep) Run(operation internal.Operation, log l
 		log.Debugf("Cluster configuration already created, skipping")
 		return operation, 0, nil
 	}
+	log.Infof("input creator is %v", operation.InputCreator)
```
Comment (piotrmiskiewicz) [inline components/kyma-environment-broker/internal/process/provisioning/create_cluster_configuration.go:43]: not needed

**Comment by ralikio** [comment]
Discredited by #2046.

**Pull request closed without merging**